notebooks/Binary Classification with the UCI Credit-card Default Dataset.ipynb

```diff
@@ -0,0 +1,628 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We attempt to mitigate the disparities in the `lightgbm` prediction using the Fairlearn postprocessing algorithm `ThresholdOptimizer`. This algorithm finds a suitable threshold for the scores (class probabilities) produced by the `lightgbm` model by optimizing the accuracy rate under the constraint that the equalized odds difference (on training data) is zero. Since our goal is to optimize balanced accuracy, we resample the training data to have the same number of positive and negative examples. (This means that `ThresholdOptimizer` is effectively optimizing balanced accuracy on the original data.)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "postprocess_est = ThresholdOptimizer(\n",
+    "    estimator=model,\n",
+    "    constraints=\"equalized_odds\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Balanced data set is obtained by sampling the same number of points from the majority class (Y=0)\n",
+    "# as there are points in the minority class (Y=1)\n",
+    "balanced_idx1 = df_train[Y_train==1].index\n",
+    "pp_train_idx = balanced_idx1.union(Y_train[Y_train==0].sample(n=balanced_idx1.size, random_state=1234).index)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df_train_balanced = df_train.loc[pp_train_idx, :]\n",
+    "Y_train_balanced = Y_train.loc[pp_train_idx]\n",
+    "A_train_balanced = A_train.loc[pp_train_idx]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "postprocess_est.fit(df_train_balanced, Y_train_balanced, sensitive_features=A_train_balanced)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "postprocess_preds = postprocess_est.predict(df_test, sensitive_features=A_test)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "models_dict = {\"Unmitigated\": (test_preds, test_scores),\n",
+    "              \"ThresholdOptimizer\": (postprocess_preds, postprocess_preds)}\n",
+    "get_metrics_df(models_dict, Y_test, A_str_test)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note that the `ThresholdOptimizer` method significantly reduces the disparity according to multiple metrics. However, note that the overall balanced error rate as well as the overall AUC get worse. Also note that since `ThresholdOptimizer` produces 0/1 predictions, the balanced error rate difference is equal to the AUC difference, and the overall balanced error rate is equal to 1 - overall AUC.\n",
```

Comment:
This gives several things to 'note' … do we have an explanation as to _why_ ?

Reply:
addressed